docs/melange-book/celsius-converter-exception.md

```diff
@@ -0,0 +1,333 @@
+of `2`. It's not possible to pass in the value of a labeled argument without
+using the `~label=value` syntax. We'll see more of labeled arguments in the
+following chapters when we introduce props (todo).
+
+You might have noticed that the function chain feeding the switch expression got
+a bit shorter, from
+
+<!--#prelude#
+let celsius = "1";
+let convert = x => x;
+let _ =
+-->
+```reasonml
+celsius |> float_of_string |> convert |> string_of_float
+```
+
+to
+
+<!--#prelude#
+let celsius = "1";
+let convert = x => x;
+let _ =
+-->
+```reasonml
+celsius |> float_of_string |> convert
+```
+
+This happened because `string_of_float`, which takes a single argument, was
+replaced by `Js.Float.toFixedWithPrecision`, which takes two arguments, and
+functions chained using `|>` can only take a single argument. But this
+one-argument restriction actually doesn't prevent us from putting
+`Js.Float.toFixedWithPrecision` in the chain! We can take advantage of OCaml's
+[partial
+application](https://reasonml.github.io/docs/en/function#partial-application)
+feature to create a one-argument function by writing
+`Js.Float.toFixedWithPrecision(~digits=2)`. Then our switch expression becomes
+
+<!--#prelude#
+let celsius = "1";
+let convert = x => x;
+let _ =
+-->
+```reasonml
+switch (celsius |> float_of_string |> convert |> Js.Float.toFixedWithPrecision(~digits=2)) {
+| exception _ => "error"
+| fahrenheit => fahrenheit ++ {js| °F|js}
+}
+```
+
+We have a working component now, but catching exceptions isn't The OCaml Way! In
+the next chapter, you'll see how to rewrite the logic using `option`.
+
+## Exercises
+
+<b>1.</b> Try changing `{js|°C = |js}` to `"°C = "`. What happens?
+
+<b>2.</b> It's possible to rewrite the `onChange` callback so that it contains a
+single expression:
+
+<!--#prelude#
+let getValueFromEvent = (evt): string => ReactEvent.Form.target(evt)##value;
+let (celsius, setCelsius) = React.useState(() => "");
+let _ =
+-->
+```reasonml
+<input value=celsius onChange={evt => setCelsius(_ => getValueFromEvent(evt))} />
+```
+
+This compiles, but it now contains a hidden bug. Do you know what silent error
+might occur?
+
+<b>3.</b> It's possible to use partial application with most functions in OCaml,
+even operators. Take a look at the following program:
+
+```reasonml
+let addFive = (+)(5);
+Js.log(addFive(2));
+Js.log(addFive(7));
+Js.log(addFive(10));
+```
+
+What do you think it outputs? Run it in [Melange
+Playground](https://melange.re/v2.0.0/playground) to confirm your hypothesis.
+
+<b>4.</b> Use the pipe last operator (`|>`) and partial application to write a
+function that takes an integer
+argument `x`, subtracts `x` from 10, and converts that result to binary. Hint:
+Use the
+[Js.Int.toStringWithRadix](https://melange.re/v2.0.0/api/re/melange/Js/Int/#val-toStringWithRadix)
+function.
+
+## Overview
+
+- Polymorphic objects (`{_.. }`) can have fields of any name and type.
```

Comment:
Same comment about this signature, not sure why there is an underscore.

Reply:
Changed this to `Js.t({..})` in https://github.com/melange-re/melange-re.github.io/pull/93/commits/0dfdf62d19e2d78beb2e40ed44045e2bab8633d9